verboseTree-maven-plugin/src/main/java/org/apache/maven/dependency/graph/SerializeGraph.java

```diff
@@ -0,0 +1,122 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.apache.maven.dependency.graph;
+
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.graph.DependencyNode;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Class to parse dependency graph and output in text format for end user to review
+ *
+ */
+public class SerializeGraph
+{
+    // will be injected eventually
+    private String outputType;
+
+    private final Map<DependencyNode, Boolean> visitedNodes;
+    private final Set<String> artifactSet;
+    private StringBuilder builder;
+
+    public SerializeGraph()
+    {
+        visitedNodes = new IdentityHashMap<DependencyNode, Boolean>( 512 );
+        artifactSet = new HashSet<String>();
+        builder = new StringBuilder();
+    }
+
+    public String serialize( DependencyNode root )
+    {
+        return dfs( root, "" ).toString();
+    }
+
+    private static void appendDependency( StringBuilder builder, DependencyNode node )
+    {
+        String scope = node.getDependency().getScope();
+        builder.append( getCoordinateString( node.getArtifact() ) );
+
+        if ( scope != null && !scope.isEmpty() )
+        {
+            builder.append( ":" ).append( scope );
+        }
+    }
+
+    private static String getCoordinateString( Artifact artifact )
+    {
+        return artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" +
+                artifact.getExtension() + ":" + artifact.getVersion();
+    }
+
+    private boolean isDuplicateArtifact( Artifact artifact )
+    {
+        String coordinateString = getCoordinateString( artifact );
+        return artifactSet.contains( coordinateString );
+    }
+
+    public StringBuilder dfs( DependencyNode node, String start )
+    {
+        builder.append( start );
+        appendDependency( builder, node );
+
+        if ( visitedNodes.containsKey( node ) )
+        {
+            builder.append( " (Omitted due to cycle.)" ).append( System.lineSeparator() );
+        }
+        else if ( isDuplicateArtifact( node.getArtifact() ) )
+        {
+            builder.append( " (Omitted due to duplicate artifact.)" ).append( System.lineSeparator() );
+        }
+        else if ( node.getDependency().isOptional() )
+        {
+            builder.append( " (Omitted due to optional dependency.)" ).append( System.lineSeparator() );
+        }
+        else
+        {
+            artifactSet.add( getCoordinateString( node.getArtifact() ) );
+            builder.append( System.lineSeparator() );
+            visitedNodes.put( node, true );
+
+            for ( int i = 0; i < node.getChildren().size(); i++ )
+            {
+                if ( start.endsWith( "+- " ) )
```

Comment:
Checking hardcoded string tends to be not future-proof -- as soon as those string changes, your logic will break. Is there any other way for you to tell where the dependency is? If not, can you make this logic more dynamic?

Reply:
Is this an issue since it's that methods responsibility to decide what strings to use anyways, they can't be changed externally? I will change the method to private to reflect that and I'll move the common strings into a variable to make maintaining it easier. Let me know if this isn't good enough because it should be fairly easy to get rid of the string comparison completely.